final/scripts/app.js

```diff
@@ -188,24 +192,59 @@
       });
     }
     // Fetch the latest data.
-    var request = new XMLHttpRequest();
-    request.onreadystatechange = function() {
-      if (request.readyState === XMLHttpRequest.DONE) {
-        if (request.status === 200) {
-          var response = JSON.parse(request.response);
-          var results = response.query.results;
+    var method = 'GET';
+    var app_id = ''; // TODO add app_id
+    var consumer_key = ''; // TODO add consumer_key
+    var consumer_secret = ''; // TODO add consumer_secret
+    var concat = '&';
+    var query = {'woeid': key, 'format': 'json'};
+    var oauth = {
+        'oauth_consumer_key': consumer_key,
+        'oauth_nonce': Math.random().toString(36).substring(2),
+        'oauth_signature_method': 'HMAC-SHA1',
+        'oauth_timestamp': parseInt(new Date().getTime() / 1000).toString(),
+        'oauth_version': '1.0'
+    };
+
+    var merged = {}; 
+    $.extend(merged, query, oauth);
+    // Note the sorting here is required
+    var merged_arr = Object.keys(merged).sort().map(function(k) {
+      return [k + '=' + encodeURIComponent(merged[k])];
+    });
+    var signature_base_str = method
+      + concat + encodeURIComponent(url)
+      + concat + encodeURIComponent(merged_arr.join(concat));
+
+    var composite_key = encodeURIComponent(consumer_secret) + concat;
+    var hash = CryptoJS.HmacSHA1(signature_base_str, composite_key);
+    var signature = hash.toString(CryptoJS.enc.Base64);
+
+    oauth['oauth_signature'] = signature;
+    var auth_header = 'OAuth ' + Object.keys(oauth).map(function(k) {
+      return [k + '="' + oauth[k] + '"'];
+    }).join(',');
+
+    $.ajax({
+      url: url + '?' + $.param(query),
+      headers: {
+        'Authorization': auth_header,
+        'Yahoo-App-Id': app_id 
+      },
+      method: 'GET',
+      success: function(data){
+        if (data) {
+          var results = data;
           results.key = key;
           results.label = label;
-          results.created = response.query.created;
+          results.created = oauth.oauth_timestamp;
           app.updateForecastCard(results);
+        } else {
+          // Return the initial weather forecast since no data is available.
+          app.updateForecastCard(initialWeatherForecast);    
         }
-      } else {
-        // Return the initial weather forecast since no data is available.
-        app.updateForecastCard(initialWeatherForecast);
       }
-    };
-    request.open('GET', url);
-    request.send();
+    });
```

Comment:
Hey man, I think that we need:

```
error: function(error) {
  app.updateForecastCard(initialWeatherForecast);
}
```

To continue working when ajax request failed.